Supervised_learning/resnet_lib.py

```diff
@@ -0,0 +1,88 @@
+import tensorflow as tf
+from tensorflow.keras.applications.resnet50 import ResNet50
+from tensorflow.keras.preprocessing import image
+from tensorflow.keras.models import Model
+from tensorflow.keras.layers import Dense, GlobalAveragePooling2D, Input, Dropout
+from tensorflow.keras.callbacks import Callback
+from tensorflow.keras.optimizers import SGD, RMSprop
+import numpy as np
+
+tf.enable_eager_execution()
+
+# Read in tfrecord files and construct dataset from it
+def create_dataset(filepath):
+    raw_image_dataset = tf.data.TFRecordDataset(filepath)
+
+    # Create a dictionary describing the features.
+    image_feature_description = {
+        'label': tf.io.FixedLenFeature([], tf.int64),
+        'image': tf.io.FixedLenFeature([], tf.string),
+    }
+
+    def _parse_image_function(example_proto):
+      # Parse the input tf.Example proto using the dictionary above.
+      return tf.io.parse_single_example(example_proto, image_feature_description)
+
+    parsed_image_dataset = raw_image_dataset.map(_parse_image_function)
+    
+    image, label = [], []
+    
+    for image_features in parsed_image_dataset:
+        label.append(image_features['label'])
+        image.append(tf.image.decode_image(image_features['image']).numpy())
+    
+    return np.asarray(image), np.asarray(label)
+
+
+# Callback function to check test set accuracy and loss for every training epoch.
+class test_callback(Callback):
+    def __init__(self, test_data):
+        self.test_data = test_data
+
+    def on_epoch_end(self, epoch, logs={}):
+        image, label = self.test_data
+        loss, acc = self.model.evaluate(image, label, verbose=0)
+        print('\nTesting loss: {}, acc: {}\n'.format(loss, acc))
+
+
+# Create transfer learning model from pretrained Resnet50 model.
+def create_model(image_size, num_classes, lr):
+    input_tensor = Input(shape=(image_size, image_size, 3))
+    # Create the base pre-trained model
+    base_model = ResNet50(weights='imagenet', include_top=False)
```

Comment:
Does ResNet50 do any image preproccessing? For example, maybe look here:

https://www.tensorflow.org/api_docs/python/tf/keras/applications/resnet/preprocess_input

Reply:
Thanks for pointing it out. Yes, I should add a data preproccessing layer. I thought it was only for imagenet inputs, but after some searching, it does help for fine tuning the model.
I've added the preprocessing unit to the model. I get similar results with and without the preprocessing unit. I will train on a few more trails to see if there is a difference.

Reply:
If you're starting from pretrained weights, then I think it is for sure needed. If training from scratch, it may not be as impactful.